File: Android/BarkCloud.Android/app/src/main/java/com/barkfluff/BarkCloud/ui/albums/AlbumsViewModel.kt
package com.barkfluff.BarkCloud.ui.albums

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.CreationExtras
import com.barkfluff.BarkCloud.BarkCloudApplication
import com.barkfluff.BarkCloud.data.cloud.AlbumCard
import com.barkfluff.BarkCloud.data.cloud.AlbumRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class AlbumsUiState(
    val isLoading: Boolean = true,
    val albums: List<AlbumCard> = emptyList(),
    val isLoadingMore: Boolean = false,
    val canLoadMore: Boolean = true,
    val isRefreshing: Boolean = false,
    val snackbar: String? = null,
)

class AlbumsViewModel(
    private val albumRepository: AlbumRepository,
) : ViewModel() {

    private val _state = MutableStateFlow(AlbumsUiState())
    val state: StateFlow<AlbumsUiState> = _state.asStateFlow()

    private var cursorUpdatedAt: Long? = null
    private var cursorAlbumId: String = ""

    fun loadIfNeeded() {
        if (_state.value.albums.isEmpty() && _state.value.isLoading) reload()
    }

    fun reload() {
        cursorUpdatedAt = null
        cursorAlbumId = ""
        viewModelScope.launch {
            _state.update { it.copy(isRefreshing = it.albums.isNotEmpty(), isLoading = it.albums.isEmpty()) }
            try {
                val page = albumRepository.listAlbums(limit = 50)
                cursorUpdatedAt = page.nextCursorUpdatedAtMillis
                cursorAlbumId = page.nextCursorAlbumId
                _state.update {
                    it.copy(isLoading = false, isRefreshing = false, albums = page.albums, canLoadMore = page.hasMore)
                }
            } catch (e: Exception) {
                _state.update { it.copy(isLoading = false, isRefreshing = false, snackbar = e.message) }
            }
        }
    }

    fun loadMore() {
        val s = _state.value
        if (!s.canLoadMore || s.isLoadingMore || s.isLoading) return
        _state.update { it.copy(isLoadingMore = true) }
        viewModelScope.launch {
            try {
                val page = albumRepository.listAlbums(
                    limit = 50,
                    cursorUpdatedAtMillis = cursorUpdatedAt,
                    cursorAlbumId = cursorAlbumId,
                )
                cursorUpdatedAt = page.nextCursorUpdatedAtMillis
                cursorAlbumId = page.nextCursorAlbumId
                _state.update {
                    it.copy(isLoadingMore = false, albums = it.albums + page.albums, canLoadMore = page.hasMore)
                }
            } catch (e: Exception) {
                _state.update { it.copy(isLoadingMore = false, snackbar = e.message) }
            }
        }
    }

    fun create(name: String) {
        if (name.isBlank()) return
        viewModelScope.launch {
            try {
                albumRepository.createAlbum(name.trim())
                reload()
            } catch (e: Exception) {
                _state.update { it.copy(snackbar = e.message) }
            }
        }
    }

    fun snackbarShown() = _state.update { it.copy(snackbar = null) }

    companion object {
        fun factory(): ViewModelProvider.Factory = object : ViewModelProvider.Factory {
            @Suppress("UNCHECKED_CAST")
            override fun <T : ViewModel> create(modelClass: Class<T>, extras: CreationExtras): T {
                val app = extras[ViewModelProvider.AndroidViewModelFactory.APPLICATION_KEY]
                    as BarkCloudApplication
                return AlbumsViewModel(app.albumRepository) as T
            }
        }
    }
}

File: Ios/BarkCloud/BarkCloud/Features/Media/Albums/AlbumsViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class AlbumsViewModel {
    struct UiState {
        var albums: [AlbumCard] = []
        var isLoading = true
        var isLoadingMore = false
        var canLoadMore = false
        var snackbar: String?

        fileprivate var cursorUpdatedAt: Date?
        fileprivate var cursorAlbumID: String = ""
    }

    var state = UiState()

    private let albums: AlbumRepository
    private var didLoad = false

    init(albums: AlbumRepository) { self.albums = albums }

    func loadIfNeeded() async {
        guard !didLoad else { return }
        didLoad = true
        await reload()
    }

    /// Перезагрузка списка. `showSpinner` поднимает полноэкранный `isLoading`
    /// (первый показ / после создания). При pull-to-refresh передаём `false` —
    /// иначе ветка `if isLoading` свернёт `ScrollView` (носитель `.refreshable`)
    /// в `ProgressView`, SwiftUI отменит задачу обновления, и gRPC-запрос упадёт
    /// с «the transport threw an unexpected error».
    func reload(showSpinner: Bool = true) async {
        if showSpinner { state.isLoading = true }
        do {
            let page = try await albums.listAlbums(limit: 50)
            state.albums = page.albums
            state.cursorUpdatedAt = page.nextCursorUpdatedAt
            state.cursorAlbumID = page.nextCursorAlbumID
            state.canLoadMore = page.hasMore
        } catch {
            state.snackbar = domainErrorMessage(error)
        }
        state.isLoading = false
    }

    func loadMoreIfNeeded(current album: AlbumCard) async {
        guard state.canLoadMore, !state.isLoadingMore, album.id == state.albums.last?.id else { return }
        state.isLoadingMore = true
        do {
            let page = try await albums.listAlbums(
                limit: 50, cursorUpdatedAt: state.cursorUpdatedAt, cursorAlbumID: state.cursorAlbumID
            )
            state.albums.append(contentsOf: page.albums)
            state.cursorUpdatedAt = page.nextCursorUpdatedAt
            state.cursorAlbumID = page.nextCursorAlbumID
            state.canLoadMore = page.hasMore
        } catch {
            state.snackbar = domainErrorMessage(error)
        }
        state.isLoadingMore = false
    }

    func create(name: String) async {
        let trimmed = name.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return }
        do {
            _ = try await albums.createAlbum(name: trimmed)
            await reload()
        } catch {
            state.snackbar = domainErrorMessage(error)
        }
    }

    func snackbarShown() { state.snackbar = nil }
}
